File: HW1/task1.py
from pyspark import SparkContext
import json
import sys


def part_a(lines):
    total_reviews = lines.count()
    print("The total number of reviews is " + str(total_reviews))
    return total_reviews


def part_b(lines, y):
    year_filter = lines.filter(lambda s: y in json.loads(s)['date'])
    year_count = year_filter.count()
    print("The number of reviews in " + y + " is " + str(year_count))
    return year_count


def part_c_d(lines, m):
    # Part C
    users = lines.map(lambda s: (json.loads(s)['user_id'], 1))
    users_group = users.groupByKey()
    users_count = users_group.count()

    # Part D
    top_users = users.reduceByKey(lambda a, b: a + b)
    top_users_col = top_users.collect()
    top_users_col = sorted(top_users_col, key=lambda x: -x[1])

    top_m = top_users_col[0:m]
    top_m_list = [[str(elem[0]), elem[1]] for elem in top_m]  # Convert to list format from tuple
    print("The number of unique users is : " + str(users_count))
    print("Top " + str(m) + " users: " + str(top_m_list[0:10]))
    return users_count, top_m_list


def part_e(lines, n):
    stopwords_file = sys.argv[3]
    with open(stopwords_file) as file:
        stopwords = file.read().splitlines()

    # process each input line by first loading, then remove punctuation/to lower case, finally remove stopwords
    reviews = lines.map(lambda line: json.loads(line)['text'])\
        .map(lambda line: line.translate({ord(i): None for i in '([,.!?:;])'}).lower())\
        .map(lambda line: " ".join([word for word in line.split() if word not in stopwords]))

    # Regular spark Word count https://spark.apache.org/examples.html
    counts = reviews.flatMap(lambda line: line.split(" ")) \
        .map(lambda word: (word, 1)) \
        .reduceByKey(lambda a, b: a + b)
    count_result = counts.collect()

    # Remove the empty string entry for edge case
    for i in range(len(count_result) - 1):
        if count_result[i][0] == "":
            del count_result[i]
            break

    count_result = sorted(count_result, key=lambda x: -x[1])
    top_n = count_result[0:n]
    top_n_list = [str(elem[0]) for elem in top_n]  # Convert to list format from tuple
    print("Top " + str(n) + "list is : " + str(top_n_list))
    return top_n_list


if __name__ == '__main__':
    """
    Run command: spark-submit task1.py $ASNLIB/public_data/review.json output.txt $ASNLIB/public_data/stopwords 2011 10 10 
    sys.argv: 1 = input_file 2 = output_file, 3 = stopwords, 4 = y, 5 = m, 6 = n
    """
    sc = SparkContext(master="local[6]", appName="task1")
    sc.setLogLevel("WARN")
    lines = sc.textFile(sys.argv[1])
    answer = dict()
    answer['A'] = part_a(lines)
    y = str(sys.argv[4])
    answer['B'] = part_b(lines, y=y)
    answer['C'], answer['D'] = part_c_d(lines, m=int(sys.argv[5]))
    answer['E'] = part_e(lines, n=int(sys.argv[6]))

    # Write answers to file
    file = open(sys.argv[2], "w")
    file.write(json.dumps(answer))
    file.close()

    print("Final answers:")
    print(json.dumps(answer))
    exit()

File: HW1/task1.scala
import org.apache.spark.rdd.RDD
import org.apache.spark.sql.SparkSession
import org.json4s._
import org.json4s.jackson.JsonMethods._
import org.json4s.JsonDSL._
import java.io.{File, FileWriter}
import scala.io.Source


object task1 {
  def part_a(lines: RDD[String]): Long = {
    val total_reviews = lines.count()
    println("The total number of reviews is " + total_reviews.toString())
    return total_reviews
  }

  def part_b(lines: RDD[String], y: String): Long = {
    val year_filter = lines.filter((s: String) => compact(parse(s) \ "date").contains(y))
    val year_count = year_filter.count()
    println("The number of reviews in " + y + " is " + year_count.toString())
    return year_count
  }

  def part_c(lines: RDD[String]): Long = {
    val users = lines.map((s: String) => (compact(parse(s) \ "user_id"), 1))
    val users_group = users.groupByKey()
    val users_count: Long = users_group.count()
    return users_count
  }

  def part_d(lines: RDD[String], m: Int): JArray = {
    val users = lines.map((s: String) => (compact(parse(s)\"user_id"), 1))
    val top_users = users.reduceByKey((a: Int, b: Int) => a + b)
    var top_users_col = top_users.collect()
    top_users_col = top_users_col.sortWith(_._2 > _._2)

    val top_m = top_users_col.slice(0, m)

    // Convert output to Json4s JArray format
    var top_m_list: List[JArray] = List()
    for (i <- top_m) {
      top_m_list :+= JArray(List(JString(i._1.replace("\"", "")), i._2))
    }
    val top_m_list1 = JArray(top_m_list)
    return top_m_list1
  }

  def part_e(lines:RDD[String], n: Int, stopwords_file: String): JArray = {
    val stopwords = Source.fromFile(stopwords_file).getLines.toList

    val reviews = (lines.map((s: String) => compact(parse(s) \ "text"))
      .map((line: String) => line.filter(!"([,.!?:;])".contains(_)).toLowerCase()))
    // Remove stopwords by filter as opposed to python

    val counts = (reviews.flatMap((line: String) => line.split(" "))
      .map((word: String) => (word, 1))
      .filter((x: (String, Int)) => !stopwords.contains(x._1) && x._1 != "")
      .reduceByKey((a: Int, b: Int) => a + b)
        .sortBy((x: (String, Int)) => -x._2))
    val count_result = counts.take(n)

    // Convert output to Json4s JArray format
    var top_n_list: List[JString] = List()
    for (i <- count_result) {
      top_n_list :+= JString(i._1.replace("\"", ""))
    }
    val top_n_list1 = JArray(top_n_list)
    return top_n_list1
  }

  def main(args: Array[String]) {
    val ss = SparkSession.builder().appName("task1").config("spark.master", "local[*]").getOrCreate()
    val sc = ss.sparkContext
    sc.setLogLevel("ERROR")
    val m = args(4).toInt
    val output = args(1)
    val y = args(3).toString
    val stopwords_file = args(2)
    val n = args(5).toInt

    val lines = sc.textFile(args(0))
    var answer: JObject = (("A", part_a(lines)) ~ ("B", part_b(lines, y))
      ~ ("C", part_c(lines)) ~ ("D", part_d(lines, m))) ~ ("E", part_e(lines, n, stopwords_file))
    println("Final answer:" + compact(answer))

    val file = new File(output)
    val bw = new FileWriter(file)
    bw.write(compact(answer))
    bw.close()
  }
}
